### PR TITLE
Feature

### PR DESCRIPTION
**Pull request by certilremy**
Added All requirement

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice progress @certilremy :+1: 
Your project looks good. I tested your methods against some edge cases and we can fix a few things before merging.
- [ ] The project should have a descriptive README.md ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_readme.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
- [ ] Pull Request should have a meaningful title and a short description ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_prs.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
- [ ] Make your commits more descriptive ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_commits.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
#### Read more of my comments for suggestions

### Martin Afani

File: lib/main.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
+module Enumerable
+  def my_each
+    element = 0
+    while element < length
+      yield(self[element])
+      element += 1
+    end
+    self
+  end
```
Comment (whiz25) [inline lib/main.rb]: - [ ] should be identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an Enumerator if no block is given
You can achieve that by doing something like this;
```ruby
def my_each
    if block_given?
      .....
    else
      return to_enum
end

File: lib/main.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  def my_each_with_index
+    element = 0
+    while element < length
+      yield(self[element], element)
+      element += 1
+    end
+  end
```
Comment (whiz25) [inline lib/main.rb]: - [ ] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ]  my_each_with_index returns an enumerator if no block is given
  ```ruby 
  array = [1,2,5,8,7,98,4]
  p array.my_each_with_index.class # not Enumerator

File: lib/main.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+  def my_select
+    new_array = []
+    my_each do |a|
+      yield(a) ? new_array.push(a) : new_array
+    end
+  end
```
Comment (whiz25) [inline lib/main.rb]: - [ ] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] my_select returns an array containing all elements of enum for which the given block returns a true value
  ```ruby
  block = proc {|x| x>4}
  array = [1,2,5,8,7,98,4]
  p array.my_select(&block) == array.select(&block) # false
- [ ] my_select returns an enumerator if no block is given
  ```ruby
  array = [1,2,5,8,7,98,4]
  p array.my_select.class # not Enumerator

File: lib/main.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  def my_all?
+    my_each do |a|
+      return false unless yield(a)
+    end
+    true
+  end
```
Comment (whiz25) [inline lib/main.rb]: - [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ]  my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
  ```ruby
  array = [1,2,5,4,7]
  p array.my_all?(Integer) == array.all?(Integer) # false
- [ ] my_all? when no block or argument is given returns true when none of the collection members are false or nil
  ```ruby
  array = [1,2,5,4,7]
  p array.my_all? == array.all? # false
- [ ] my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
  ```ruby
  array = %w[dog door rod blade]
  p array.my_all?(/c/) == array.all?(/c/) # false
- [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
  ```ruby
  array = %w[dog door rod blade]
  p array.my_all?(5) == array.all?(5) # false

File: lib/main.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  def my_any?
+    exist = false
+    my_each do |element|
+      exist = yield(element)
+      break if exist
+    end
+    exist
+  end
```
Comment (whiz25) [inline lib/main.rb]: - [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
 ```ruby 
  array = [nil, false, true, []]
  p array.my_any? == array.any? # false
```
- [ ]  when a Regex is passed as an argument returns false if none of the collection matches the Regex
  ```ruby
  array = %w[dog door rod blade]
  p array.my_any?(/c/) == array.any?(/c/) # false
  ```
 - [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
  ```ruby
  array = %w[dog door rod blade]
  p array.my_any?(5) == array.any?(5) # false
 ```

File: lib/main.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+  def my_none?
+    (my_any? { |i| yield(i) == true }) != true
+  end
```
Comment (whiz25) [inline lib/main.rb]: - [x] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] my_none? when no block or argument is given returns true only if none of the collection members is true
```ruby
  array = [1,2,5,4,7]
  p array.my_none? == array.none? # false
```
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
  array = [1,2,5,4,7]
  p array.my_none?(Integer) == array.none?(Integer) # false
```
- [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
  array = %w[dog door rod blade]
  p array.my_none?(/c/) == array.none?(/c/) # false
```
 - [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
  array = %w[dog door rod blade]
  p array.my_none?(5) == array.none?(5) # false
```

File: lib/main.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      my_each do |element|
+        new_arr << proc.call(element)
+      end
+    elsif proc.nil? && block_given?
+      my_each do |element|
+        new_arr << yield(element)
+      end
+    end
+    new_arr
+  end
```
Comment (whiz25) [inline lib/main.rb]: - [x] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an Enumerator if no block is given
```ruby
 array = [1,5,8,7,9,5,5,55,5]
 p array.my_map #error
 LocalJumpError:
     no block given (yield)
```

File: lib/main.rb [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+  def my_inject(s_t)
+    my_each { |i| s_t = yield(s_t, i) }
+    s_t
+  end
+end
```
Comment (whiz25) [inline lib/main.rb]: - [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] my_inject combines all elements of enum by applying a binary operation, specified by a block
```ruby
  range = Range.new(5, 50)
  p range.my_inject(4) { |prod, n| prod * n } # returns this error
  # NoMethodError:
     #  undefined method `length' for 5..50:Range
    # ./enumerables.rb:76:in `my_inject'
```
- [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
  array = [1,7,8,89,9,6,5,12,4,2,5]
  p array.my_inject(:+) == array.inject(:+) # return this error
  # LocalJumpError:
     # no block given (yield)
    # ./enumerables.rb:78:in `block in my_inject'
    # ./enumerables.rb:76:in `each'
    # ./enumerables.rb:76:in `my_inject'
```

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Hey @certilremy 
I know you have made some changes that are good but your methods still fail against the same edge cases I pointed out in my previous review.
Check my suggestions again from the previous review. Take your time and make sure everything works well before submitting it for another review.

### Martin Afani

**Comment by certilremy** [comment]
> ## Hey @certilremy
> I know you have made some changes that are good but your methods still fail against the same edge cases I pointed out in my previous review.
> Check my suggestions again from the previous review. Take your time and make sure everything works well before submitting it for another review.
> 
> ### Martin Afani

Thank you for your feedback. I didn't quite understand what you need me to do. I read and read the project from the Odin project I can't see it.

File: lib/main.rb [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+  def my_select
+    return enum_for(:my_select) unless block_given?
+
+    new_array = []
+
+    my_each do |a|
+      yield(a) ? new_array.push(a) : new_array
+    end
+  end
```
Comment (suretrust) [inline lib/main.rb]: * [ ] should have working `my_select` **(covered by automated tests)** identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [ ] enumerables my_select returns an array containing all elements of enum for which the given block returns a true value
```ruby
[1, 2, 3, 4, 5, 6, 7, 8, 9].select { |num| num < (1 + 9) / 2 } != [1, 2, 3, 4, 5, 6, 7, 8, 9].my_select { |num| num < (1 + 9) / 2 }
```

File: lib/main.rb [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+      end
+    elsif block_given?
+      my_each do |element|
+        return false unless yield(element)
+      end
+    else
+      my_each do |element|
+        return false unless element
+      end
+    end
```
Comment (suretrust) [inline lib/main.rb]: * [ ] should have working `my_any?` **(covered by automated tests)** identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- [ ] enumerables my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
[0,1,2,3,4,5,6,7,8,9].my_all?(3) != [0,1,2,3,4,5,6,7,8,9].all?(3)
```

File: lib/main.rb [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        return true if yield(element)
+      end
+    else
+      my_each do |element|
+        return true if element
+      end
+    end
+
+    false
+  end
```
Comment (suretrust) [inline lib/main.rb]: * [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- [ ] enumerables my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
['saheed','oladele','suretrust'].my_any?('cat') != ['saheed','oladele','suretrust'].any?('cat')
```

File: lib/main.rb [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        return false if yield(element)
+      end
+    else
+      my_each do |element|
+        return false if element
+      end
+    end
+
+    true
+  end
```
Comment (suretrust) [inline lib/main.rb]: * [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- [ ] enumerables my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
['saheed','oladele','suretrust'].my_none?(5) != ['saheed','oladele','suretrust'].none?(5)
```

File: lib/main.rb [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+  def my_count(number = nil)
+    return total = lenght unless block_given?
```
Comment (suretrust) [inline lib/main.rb]: `length` typo-ed as `lenght`

File: lib/main.rb [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+      return arr[-1] if sym.to_s == '='
+
+      arr.my_each do |element|
+        result = result.send(sym.to_s, element)
+      end
+    else
+      arr.my_each do |element|
+        result = yield(result, element)
+      end
+    end
```
Comment (suretrust) [inline lib/main.rb]: * [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- [ ] Enumerables my_inject combines all elements of enum by applying a binary operation, specified by a block.
```ruby
[1,2,3,4,5].my_inject(&:+) != [1,2,3,4,5].inject(&:+)
```

- [ ] Enumerables my_inject, when a symbol is specified, combines each element of the collection by applying the symbol as a named method.
```ruby
[1,2,3,4,5].my_inject(:+) != [1,2,3,4,5].inject(:+)
```

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @certilremy, :wave: 

### Great job thus far! :raised_hands: Please see my comments on your code. :point_down: 

**Kindly submit for another code review once these have been fixed.** _Happy coding!_ :tada:

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good Progress @certilremy :+1: 
You have made a tremendous progress in this project. Let's fix the remaining failing edge cases so we can merge. 
- [ ] my_select returns an array containing all elements of enum for which the given block returns a true value
- [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [ ] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [ ] my_count counts the number of items in enum that are equal to item if an argument is given
- [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
rspec ./spec/test_spec.rb:215 # enumerables 
- [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

### Martin Afani

**Comment by whiz25** [comment]
> > ## Hey @certilremy
> > I know you have made some changes that are good but your methods still fail against the same edge cases I pointed out in my previous review.
> > Check my suggestions again from the previous review. Take your time and make sure everything works well before submitting it for another review.
> > ### Martin Afani
> 
> Thank you for your feedback. I didn't quite understand what you need me to do. I read and read the project from the Odin project I can't see it.

What I meant was that some of your methods were not behaving like the original ruby methods. For example, your `my_map` was not giving the same result as the built-in `map` method.
This is something that I pointed out in my first review of your PR. So on testing your methods again against some tests, I was still getting the same result as before even after the changes you made.
This is why I referred you to the first review.

File: lib/main.rb [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+        end
+        ind
+      else
+        'method error'
+      end
+    else
+      'block not found'
+    end
+  end
+end
```
Comment (stratospherique) [inline lib/main.rb]: * [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
    # NameError:
       undefined local variable or method `length' for 5..50:Range
     # ./lib/main.rb:121:in `my_inject'
  ```

* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    # NameError:
       undefined local variable or method `ind' for 5..50:Range
       Did you mean?  find
                      end
                      end
                      and
                      END
     # ./lib/main.rb:154:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Amazing work implementing the suggested changes :clap:  :clap: However, I got some cases in which `my_inject` doesn't behave the way it should be. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your method :nerd_face: 

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: lib/main.rb [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+      end
+    end
+    new_arr
+  end
+
+  def my_inject(given = self[0], symbol = nil)
+    if block_given?
+      given ||= 0
+      index = 1
+      while index < length
```
Comment (Forison) [inline lib/main.rb]: * [ ] length is undefined kindly rectify it.
 ```
range = Range.new(5, 50)
p range.my_inject(4) { |prod, n| prod * n }
range_two = Range.new(5, 50)
p range_two.my_inject(2, :*)
 ``` 
returns
 ```main.rb:122:in `my_inject': undefined local variable or method `length' for 5..50:Range (NameError) ```

**Review by Forison — CHANGES_REQUESTED** [review]
## Status: Require Changes🛑
## Hi @certilremy,
I can see you have come a long a way🙂, That is the spirit, coding is all bout problem-solving🙂.
The issue raised in the last review still prevails🤔, check below for comment on the error👇
# Kindly make changes and submit the project for review
## Happy coding💪
## Addo Forison

**Comment by shshamim63** [comment]
## Status :heavy_check_mark: [Approved] :heavy_check_mark: 
___
Hi @certilremy :wave: ,
Project contains descriptive readme, you have covered all the changes asked by the previous reviewer. All the methods are working properly. Awesome job :clap: . 
___
### This :+1: is from me. You can merge it. and don't forget to fill the project completion form.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

**Pull request merged**